utils/onboarding/backend_interface.py

```diff
@@ -13,87 +23,80 @@ def _headers():
 
 
 def _query_for_trace_id(trace_id, validator=None):
-    path = f"/api/v1/trace/{trace_id}"
-    host = "https://dd.datadoghq.com"
+    url = f"{API_HOST}/api/v1/trace/{trace_id}"
 
-    try:
-        r = requests.get(f"{host}{path}", headers=_headers(), timeout=10)
-        logger.debug(f" Backend response status for trace_id [{trace_id}]: [{r.status_code}]")
-        if r.status_code == 200:
-            logger.debug(f" Backend response for trace_id [{trace_id}]: [{r.text}]")
-            # Check if it's  not a old trace
-            trace_data = r.json()
-            if validator:
-                logger.info("Validating backend trace...")
-                if not validator(trace_id, trace_data):
-                    logger.info("Backend trace is not valid")
-                    return -1, None
-                logger.info("Backend trace is valid")
-            root_id = trace_data["trace"]["root_id"]
-            start_time = trace_data["trace"]["spans"][root_id]["start"]
-            start_date = datetime.fromtimestamp(start_time)
-            if (datetime.now() - start_date).days > 1:
-                logger.info("Backend trace is too old")
-                return -1, None
-            runtime_id = trace_data["trace"]["spans"][root_id]["meta"]["runtime-id"]
-            return r.status_code, runtime_id
-        return r.status_code, None
-    except Exception as e:
-        logger.error(f"Error received connecting to host: [{host}] {e} ")
-        return -1, None
+    trace_data = _make_request(url, headers=_headers())
+    if validator:
+        logger.info("Validating backend trace...")
+        if not validator(trace_id, trace_data):
+            logger.info("Backend trace is not valid")
+            return None
+        logger.info("Backend trace is valid")
 
+    root_id = trace_data["trace"]["root_id"]
+    start_time = trace_data["trace"]["spans"][root_id]["start"]
+    start_date = datetime.fromtimestamp(start_time)
+    if (datetime.now() - start_date).days > 1:
+        logger.info("Backend trace is too old")
+        return None
 
-def _query_for_profile(runtime_id):
-    path = "/api/unstable/profiles/list"
-    host = "https://dd.datadoghq.com"
+    return trace_data["trace"]["spans"][root_id]["meta"]["runtime-id"]
+
+
+def _make_request(
+    url, headers=None, method="get", json=None, request_timeout=10, retry_delay=1, backoff_factor=2, max_retries=6
+):
+    for _attempt in range(max_retries):
+        try:
+            r = requests.request(method=method, url=url, headers=headers, json=json, timeout=request_timeout)
+            logger.debug(f" Backend response status : [{r.status_code}]")
+            if r.status_code == 200:
+                return r.json()
+
+            if r.status_code == 429:
+                retry_after = _parse_retry_after(r.headers.get("X-RateLimit-Reset"))
```

Comment:
This header was found in the docs: https://docs.datadoghq.com/api/latest/rate-limits/

During testing, I found that the Datadog API does not set `Retry-After` which highlighted another bug in my code: https://gitlab.ddbuild.io/DataDog/system-tests/-/jobs/771201905. Two bugs with one test?